### PR TITLE
Update MSC3958 support to interact with intentional mentions.

### PR DESCRIPTION
**Pull request by clokep**
This updates the [MSC3958 support](https://github.com/matrix-org/matrix-spec-proposals/pull/3958) to match the latest version of the MSC. This should correspond to the [following changes](https://github.com/matrix-org/matrix-spec-proposals/pull/3958/commits/876e19fa882f544cbfde1cb026ac0b9f7ef23647):

* Update the rule ID.
* Use `event_property_is` instead of `event_match`.
* Move the rule below the rules from MSC3952 (intentional mentions) allowing for an edit which mentions a user to still notify.

### CODE REVIEW

**Comment by clokep** [comment]
I tested this by:

1. Make a room with 3 users in it
2. Enable intentional mentions (via labs on Element Web develop; I needed to lie and update the `/versions` response to include `v1.7`)
3. Enable MSC3958 in the Synapse config
4. Send a message in the room via `alice`; the count for `bob` and `charlie` go to 1
5. Edit the message; note that the count stays at 1
6. Send a new message as `alice`; the count for `bob` and `charlie` go to 2
7. Edit the message from 6 to @ `bob`; the count for `bob` is now 3 & red; the count for `charlie` stays at 2

File: rust/src/push/base_rules.rs [hunk 1]
```diff
@@ -241,6 +225,21 @@ pub const BASE_APPEND_OVERRIDE_RULES: &[PushRule] = &[
...
         default_enabled: true,
     },
+    // We don't want to notify on edits *unless* the edit directly mentions a
+    // user, which is handled above.
+    PushRule {
+        rule_id: Cow::Borrowed("global/override/.org.matrix.msc3958.suppress_edits"),
+        priority_class: 5,
+        conditions: Cow::Borrowed(&[Condition::Known(KnownCondition::EventPropertyIs(
+            EventPropertyIsCondition {
+                key: Cow::Borrowed("content.m\\.relates_to.rel_type"),
```
Comment (erikjohnston) [inline rust/src/push/base_rules.rs:235]:
```suggestion
                key: Cow::Borrowed(r"content.m\.relates_to.rel_type"),
```

Might be an easier read? (Those are called raw strings)

File: rust/src/push/base_rules.rs [hunk 2]
```diff
@@ -241,6 +225,21 @@ pub const BASE_APPEND_OVERRIDE_RULES: &[PushRule] = &[
...
+                value: Cow::Borrowed(&SimpleJsonValue::Str(Cow::Borrowed("m.replace"))),
```
Comment (erikjohnston) [inline rust/src/push/base_rules.rs:236]:
```suggestion
                value: Cow::Owned(SimpleJsonValue::Str(Cow::Borrowed("m.replace"))),
```

No?

File: rust/src/push/evaluator.rs [hunk 3]
```diff
@@ -117,7 +117,7 @@ impl PushRuleEvaluator {
         msc3931_enabled: bool,
     ) -> Result<Self, Error> {
         let body = match flattened_keys.get("content.body") {
-            Some(JsonValue::Value(SimpleJsonValue::Str(s))) => s.clone(),
+            Some(JsonValue::Value(SimpleJsonValue::Str(s))) => s.clone().to_string(),
```
Comment (erikjohnston) [inline rust/src/push/evaluator.rs:120]:
```suggestion
            Some(JsonValue::Value(SimpleJsonValue::Str(s))) => s.clone().into_owned(),
```

is probably more technically efficient (reduces it to one clone). The alternative would also be `s.to_string()`.

**Review by erikjohnston — APPROVED** [review]
Just some small nits that don't actually matter at all in the slightest

File: rust/src/push/base_rules.rs (hunk 1 above)
Comment (clokep) [reply rust/src/push/base_rules.rs:235]: That would be fine too, yeah. I didn't realize Rust had those like Python does. 👍

**Pull request merged**